### PR TITLE
Support rollingUpdate config of webserver

### PR DESCRIPTION
**Pull request by junnplus**
webserver pod waiting long time for init container complete, using one replica will make the webserver unavailable.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: chart/templates/webserver/webserver-deployment.yaml
Comment (XD-DENG) [inline chart/templates/webserver/webserver-deployment.yaml]: The original code (`maxUnavailable: 1`) already looks good to me when `replicas` >=2.

It's possible that users explicitly specify the `replicas` to 1, for some purposes. In that case, users take the responsibility and should be aware of the result.

Please correct me if I misunderstand/miss anything.

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml]: > It's possible that users explicitly specify the replicas to 1, for some purposes. In that case, users take the responsibility and should be aware of the result.

I don't think so, now replicas defaults to 1, upgrade chart always waiting long time, this will make the webserver unavailable.
Of course, set replicas to >=2 maybe can solve this problem, but everyone needs to do it.

Comment (XD-DENG) [reply chart/templates/webserver/webserver-deployment.yaml]: I'm not really convinced by this. I would like to have some valuable inputs from other committers as well @mik-laj @dimberman

Comment (mik-laj) [reply chart/templates/webserver/webserver-deployment.yaml]: It makes sense for me to provide support for service updates by default, with a guarantee of its continuous availability. I don't think HA for the webserver is often deployed because the webserver is not used very heavily because the key feature are provided by the webserver and worker. The web server only supports the use of Airflow but is an optional component. The Web Server only supports the use of Airflow, but from a design perspective it is an optional, non-essential component, so in most cases, HA is not necessary.

However, to handle both cases, we can add a condition. If `replicas > 2`, then the `maxUnavailable = replicas - 1,` If `replicas == 1`, then `maxUnavailable = 0`. What do you think about it?

Comment (Junnplus) [reply chart/templates/webserver/webserver-deployment.yaml:42]: @mik-laj Can we let the user decide `maxUnavailable` value?

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml]: Can we let users decide `maxUnavailable` ?

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml]: > Can we let users decide maxUnavailable ?

ping @mik-laj @XD-DENG

Comment (XD-DENG) [reply chart/templates/webserver/webserver-deployment.yaml]: I agree it's reasonable to let the users decide `maxUnavailable` (e.g. in `values.yaml`), **with proper default/fall-back value.**

Comment (mik-laj) [reply chart/templates/webserver/webserver-deployment.yaml]: I agree with both.

**Comment by Junnplus** [comment]
ping @XD-DENG @mik-laj

File: chart/templates/webserver/webserver-deployment.yaml
Comment (XD-DENG) [inline chart/templates/webserver/webserver-deployment.yaml:42]: maybe more straight forward to have `maxUnavailable: {{ .Values.webserver.rollingUpdate.maxUnavailable }}` here. Let me know if you have different opinion.

**Review by XD-DENG — CHANGES_REQUESTED** [review]
Thanks for the changes. 

Other than the in-line comment below, the PR title and description should be updated accordingly, after your latest change. The scope now is more like "allow configuring `rollingUpdate.maxUnavailable`" instead of "make webserver rolling update" (This is also why I didn't realise there is new change to review in this PR).

**Comment by dimberman** [comment]
I'm ok to merge this once @XD-DENG 's comments are addressed

File: chart/templates/webserver/webserver-deployment.yaml
Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml:42]: I think, It may be better to support both `maxSurge` and `maxUnavailable`.

https://github.com/apache/airflow/pull/13906/files#diff-a9d07052701ab3b718c82ac0fe74b965a8260b06bb5906362081718b0ee593b4R419

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml:42]: Like you said, `maxUnavailable` maybe more straight forward, but the PR title is "make webserver rolling update", not just `maxUnavailable`.

Comment (XD-DENG) [reply chart/templates/webserver/webserver-deployment.yaml:42]: As I mentioned above, the change you intend to make here is no longer "make webserver rolling update", instead, it's allowing more configuraitons.

As already suggested, please update the PR title and description; or create a new PR accordingly.

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml:42]: @XD-DENG Well.. I modified the title.

Comment (XD-DENG) [reply chart/templates/webserver/webserver-deployment.yaml:42]: Do you mind addressing the comment above? I don't see it necessary to use `toYaml` here.

> maybe more straight forward to have `maxUnavailable: {{ .Values.webserver.rollingUpdate.maxUnavailable }}` here. Let me know if you have different opinion.

In addition, please also update the commit subject and add description properly (that's what appear eventually in the commit history)

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml:42]: Rolling Update of kubernetes not just `maxUnavailable`

>I think, It may be better to support both maxSurge and maxUnavailable.
> https://github.com/apache/airflow/pull/13906/files#diff-a9d07052701ab3b718c82ac0fe74b965a8260b06bb5906362081718b0ee593b4R419

Comment (XD-DENG) [reply chart/templates/webserver/webserver-deployment.yaml:42]: Then where is your `maxSurge` in `values.yaml`?  :-)
As well as the necessary update in `chart/README.md`? 

Please make *all* necessary changes and rebase.

Comment (junnplus) [reply chart/templates/webserver/webserver-deployment.yaml:42]: `rollingUpdate.maxSurge` is an optional field, default value is 25%, I added the annotation of maxsurge.

**Pull request closed without merging**
